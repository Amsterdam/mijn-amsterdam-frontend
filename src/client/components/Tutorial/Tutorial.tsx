import classnames from 'classnames';
import FocusTrap from 'focus-trap-react';
import React, { CSSProperties, useEffect, useRef, useState } from 'react';
import ReactDOM from 'react-dom';
import { IconArrowRrimaryWhite } from '../../assets/icons';
import { useDetectResizing, useModalRoot, usePhoneScreen } from '../../hooks';
import { CloseButton } from '../Button/Button';
import styles from './Tutorial.module.scss';

function TutorialItem({ el }: { el: any }) {
  const heading = el.querySelector('[class^="Heading_Heading"]') || el;
  const pos = heading.getBoundingClientRect();
  const styledElement = window.getComputedStyle(heading, null);
  const fontSize = styledElement.getPropertyValue('font-size');
  const lineHeight = styledElement.getPropertyValue('line-height');
  const padding = styledElement.getPropertyValue('padding');
  const [
    text,
    fromDirection = 'right-top',
    label,
  ] = el.dataset.tutorialItem.split(';');

  const ref = useRef<HTMLDivElement | null>(null);
  const headingRef = useRef<HTMLHeadingElement | null>(null);
  const [itemPos, setItemPos] = useState({ width: 0, left: 0 });
  const [headingPos, setHeadingPos] = useState({ width: 0, left: 0 });

  useEffect(() => {
    if (headingRef.current !== null) {
      setHeadingPos((headingRef.current as any).getBoundingClientRect());
    }
    if (ref.current !== null) {
      setItemPos((ref.current as any).getBoundingClientRect());
    }
  }, [ref, headingRef]);

  const itemWidth = (itemPos as any).width;
  const headingWidth = (headingPos as any).width + pos.left;
  const left =
    fromDirection.startsWith('right') &&
    headingWidth + itemWidth > window.innerWidth
      ? headingWidth + itemWidth - window.innerWidth
      : 0;
  const textStyle: CSSProperties = {
    transform: `translateX(-${left}px)`,
  };

  return (
    <div
      className={classnames(styles.TutorialItem, styles[fromDirection])}
      style={{
        left: pos.left,
        top: pos.top + window.pageYOffset,
      }}
    >
      <h3
        ref={headingRef}
        className={styles.TutorialItemHeading}
        style={{ fontSize, lineHeight, padding }}
      >
        {label || heading.textContent}
      </h3>
      <div className={styles.TutorialItemInner}>
        <p className={styles.TutorialText} style={textStyle}>
          <span ref={ref}>{text}</span>
        </p>
        <span aria-hidden="true" className={styles.ArrowIconContainer}>
          <IconArrowRrimaryWhite className={styles.ArrowIcon} />
        </span>
      </div>
    </div>
  );
}

interface TutorialProps {
  onClose: () => void;
}

export default function Tutorial({ onClose }: TutorialProps) {
  const tutorialItems = Array.from(
    document.querySelectorAll('[data-tutorial-item]')
  );

  const isResizing = useDetectResizing();
  const isPhoneScreen = usePhoneScreen();

  // Check if positions are calculated
  return ReactDOM.createPortal(
    <FocusTrap focusTrapOptions={{ escapeDeactivates: false }}>
      <div
        className={classnames(
          styles.Tutorial,
          isResizing && !isPhoneScreen && styles.TutorialResizing
        )}
        style={{ height: document.body.clientHeight }}
      >
        {tutorialItems
          .filter((el) => {
            return !!el.getAttribute('data-tutorial-item');
          })
          .map((el, i) => (
            <TutorialItem key={i} el={el} />
          ))}
        <CloseButton
          title="Uitleg verbergen"
          onClick={onClose}
          className={styles.CloseTutorial}
        />
      </div>
    </FocusTrap>,
    useModalRoot()
  );
}
