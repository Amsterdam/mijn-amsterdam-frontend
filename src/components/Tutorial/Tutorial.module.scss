@import 'styles/_global.scss';

$tutorial-items-padding: 1rem;
$tutorial-items-radius: 1rem;

.Tutorial {
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  background-color: rgba($color-neutral-grey5, 0.8);
  z-index: z-index(tutorial-overlay);
  overflow: hidden;
  min-width: $main-min-width;
}

.TutorialItem {
  color: $color-white;
  font-size: 1.7rem;
  line-height: 2.2rem;
  font-weight: bold;
  text-align: center;
  position: relative;
}

.MyUpdatesItem {
  width: 25rem;
  margin-top: -16rem;
  margin-left: 23rem;

  svg {
    transform: rotate(180deg);
    margin-left: -10rem;
    margin-top: 1rem;
  }

  @include media-query('<=tablet') {
    margin-top: -16rem;
  }

  @include media-query('<=phone') {
    margin-top: -15rem;
    margin-left: 2rem;

    svg {
      margin-top: 2rem;
      margin-left: 16rem;
      height: 8rem;
      transform: rotate(150deg);
    }
  }
}

.MyChaptersItem {
  width: 28rem;
  margin-top: -3rem;
  margin-left: 16rem;

  svg {
    transform: rotate(180deg);
    margin-left: -6rem;
    transform: scaleX(-1);
    margin-top: 1rem;
  }

  @include media-query('<=tablet') {
    width: 28rem;
    margin-top: -3rem;
    margin-left: -26rem;

    svg {
      transform: rotate(0deg);
      margin-left: 6rem;
      margin-top: 1rem;
    }
  }

  @include media-query('<=phone') {
    margin-top: -15rem;
    margin-left: 0rem;

    svg {
      transform: rotate(170deg);
      margin-left: 14rem;
      margin-top: 2rem;
      height: 8rem;
    }
  }
}

.MyCasesItem {
  width: 30rem;
  margin-top: -17.5rem;
  margin-left: 22rem;

  svg {
    transform: rotate(-10deg) scaleX(-1);
    margin-left: 14rem;
    margin-top: 1rem;
  }

  @include media-query('<=phone') {
    margin-top: -17rem;
    margin-left: 2rem;
    width: 26rem;

    svg {
      margin-left: -24rem;
      margin-top: -2rem;
      transform: rotate(50deg) scaleY(-1);
      height: 8rem;
    }
  }
}

.MyAreaItem {
  width: 46rem;
  margin-top: -30rem;
  margin-left: -36rem;

  svg {
    margin-left: 8rem;
    transform: scaleY(-1);
    margin-top: 1rem;
  }

  @include media-query('<=phone') {
    width: 22rem;
    margin-top: -14rem;
    margin-left: 2rem;

    svg {
      margin-left: 6rem;
      transform: rotate(180deg);
      height: 8rem;
    }
  }
}

.MyTipsItem {
  width: 40rem;
  margin-top: -4rem;
  margin-left: 6rem;

  svg {
    margin-left: -9rem;
    transform: scaleX(-1);
    margin-top: 1rem;
  }

  @include media-query('<=phone') {
    width: 22rem;
    margin-top: -2rem;
    margin-left: 2rem;

    svg {
      height: 8rem;
      margin-left: 4rem;
    }
  }
}

%tutorialHeading {
  position: relative;
  z-index: z-index(tutorial-items);
  background-color: $color-white;
  padding: $tutorial-items-padding;
  border-radius: $tutorial-items-radius;
  margin-left: -$tutorial-items-padding;
}

.TutorialItems {
  [id='MyChaptersHeader'],
  [id='MyAreaHeader'],
  [id='MyTipsHeader'],
  [id='MyCasesHeader'],
  [id='MyUpdatesHeader'] {
    @extend %tutorialHeading;
  }

  [id='MyCasesHeader'] {
    display: inline-block;
    margin-top: -$tutorial-items-padding;
    margin-bottom: calc(#{$tutorial-items-padding} * 1);
  }

  [id='MyChaptersHeader'] {
    display: inline-block;
    margin-top: -$tutorial-items-padding;
    margin-bottom: $tutorial-items-padding;
  }

  [id='MyAreaHeader'] {
    display: inline-block;
    margin-top: -$tutorial-items-padding;
    margin-bottom: -$tutorial-items-padding;

    @include media-query('<=phone') {
      margin-left: 0rem;
      display: initial;
      background-position: 0rem 1.2rem;
      padding-left: 2rem;
    }
  }

  [id='MyTipsHeader'] {
    padding: $tutorial-items-padding;
    margin-top: -$tutorial-items-padding;
    margin-right: -$tutorial-items-padding;
    margin-bottom: -$tutorial-items-padding;

    @include media-query('<=phone') {
      margin-left: 0rem;
      background-position: 0rem 1.2rem;
      padding-left: 2rem;
    }
  }
}
