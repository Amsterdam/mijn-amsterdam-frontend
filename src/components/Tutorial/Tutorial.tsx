import React, { CSSProperties } from 'react';
import ReactDOM from 'react-dom';
import styles from './Tutorial.module.scss';

import { ReactComponent as ArrowIcon } from 'assets/icons/Arrow__primary-white.svg';
import useModalRoot from 'hooks/modalRoot.hook';
import classnames from 'classnames';
import useDetectResizing from 'hooks/detectResize.hook';
import { usePhoneScreen } from 'hooks/media.hook';
import { useRef, useEffect, useState } from 'react';
import FocusTrap from 'focus-trap-react';
import { CloseButton } from '../Button/Button';

interface TutorialProps {
  onClose: () => void;
}

function TutorialItem({ el }: { el: any }) {
  const heading = el.querySelector('[class^="Heading_Heading"]') || el;
  const pos = heading.getBoundingClientRect();
  const styledElement = window.getComputedStyle(heading, null);
  const fontSize = styledElement.getPropertyValue('font-size');
  const lineHeight = styledElement.getPropertyValue('line-height');
  const padding = styledElement.getPropertyValue('padding');
  const [text, fromDirection = 'right-top'] = el.dataset.tutorialItem.split(
    ';'
  );

  const ref = useRef<HTMLDivElement | null>(null);
  const headingRef = useRef<HTMLHeadingElement | null>(null);
  const [itemPos, setItemPos] = useState({ width: 0, left: 0 });
  const [headingPos, setHeadingPos] = useState({ width: 0, left: 0 });

  useEffect(() => {
    if (headingRef.current !== null) {
      setHeadingPos((headingRef.current as any).getBoundingClientRect());
    }
    if (ref.current !== null) {
      setItemPos((ref.current as any).getBoundingClientRect());
    }
  }, [ref.current, headingRef.current]);

  const itemWidth = (itemPos as any).width;
  const headingWidth = (headingPos as any).width + pos.left;
  const left =
    fromDirection.startsWith('right') &&
    headingWidth + itemWidth > window.innerWidth
      ? headingWidth + itemWidth - window.innerWidth
      : 0;
  const textStyle: CSSProperties = {
    transform: `translateX(-${left}px)`,
  };

  return (
    <div
      className={classnames(styles.TutorialItem, styles[fromDirection])}
      style={{
        left: pos.left,
        top: pos.top + window.pageYOffset,
      }}
    >
      <h3
        ref={headingRef}
        className={styles.TutorialItemHeading}
        style={{ fontSize, lineHeight, padding }}
      >
        {heading.textContent}
      </h3>
      <div className={styles.TutorialItemInner}>
        <p className={styles.TutorialText} style={textStyle}>
          <span ref={ref}>{text}</span>
        </p>
        <span aria-hidden="true" className={styles.ArrowIconContainer}>
          <ArrowIcon className={styles.ArrowIcon} />
        </span>
      </div>
    </div>
  );
}

export default function Tutorial({ onClose }: TutorialProps) {
  const tutorialItems = Array.from(
    document.querySelectorAll('[data-tutorial-item]')
  );

  const isResizing = useDetectResizing();
  const isPhoneScreen = usePhoneScreen();

  // Check if positions are calculated
  return ReactDOM.createPortal(
    <FocusTrap focusTrapOptions={{ escapeDeactivates: false }}>
      <div
        className={classnames(
          styles.Tutorial,
          isResizing && !isPhoneScreen && styles.TutorialResizing
        )}
        style={{ height: document.body.clientHeight }}
      >
        {tutorialItems.map((el, i) => (
          <TutorialItem key={i} el={el} />
        ))}
        <CloseButton
          title="Uitleg verbergen"
          onClick={onClose}
          className={styles.CloseTutorial}
        />
      </div>
    </FocusTrap>,
    useModalRoot()
  );
}
