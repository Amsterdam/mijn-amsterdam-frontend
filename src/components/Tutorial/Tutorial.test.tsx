import React from 'react';
import { mount } from 'enzyme';
import Tutorial from './Tutorial';

describe('tutorial', () => {
  it('Renders without crashing', () => {
    const comp = mount(
      <div>
        <div data-tutorial-item="Hello world!left-top">This is happening</div>
        <div data-tutorial-item="Goody two shoes!left-bottom">
          On a large scale
        </div>
        <Tutorial />
      </div>
    );

    expect(comp.html()).toMatchSnapshot();
    comp.unmount();
  });
});
