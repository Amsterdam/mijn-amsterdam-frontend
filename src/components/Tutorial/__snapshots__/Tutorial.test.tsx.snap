// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tutorial Renders without crashing 1`] = `"<div class=\\"Tutorial\\"><div class=\\"TutorialItem MyUpdatesItem\\" style=\\"top: 0px; left: 0px;\\">Hier ziet u nieuwe berichten van onze afdelingen die uw aandacht vragen<br><svg>Arrow__primary-white.svg</svg></div><div class=\\"TutorialItem MyChaptersItem\\" style=\\"top: 0px; left: 0px;\\"><svg>Arrow__primary-white.svg</svg><br>Dit zijn de onderwerpen waarover u iets heeft bij de gemeente</div><div class=\\"TutorialItem MyCasesItem\\" style=\\"top: 0px; left: 0px;\\"><svg>Arrow__primary-white.svg</svg><br>Dit is een overzicht van uw lopende aanvragen of wijzigingen</div><div class=\\"TutorialItem MyAreaItem\\" style=\\"top: 0px; left: 0px;\\">Hier ziet u informatie van de gemeente, bijvoorbeeld over afval, parkeren en bekendmakingen<br><svg>Arrow__primary-white.svg</svg></div><div class=\\"TutorialItem MyTipsItem\\" style=\\"top: 0px; left: 0px;\\"><svg>Arrow__primary-white.svg</svg><br>Hier geven wij u handige tips, bijvoorbeeld over de regelingen en voorzieningen van de gemeente</div></div>"`;
